lib/auth/webauthncli/fido2.go

```diff
@@ -592,98 +585,11 @@ func runOnFIDO2Devices(
 
 	// Run the callback again with the informed PIN.
 	// selectDevice is used since it correctly deals with cancellation.
-	_, _, err = selectDevice(ctx, pin, []deviceWithInfo{dev}, cb)
+	cb = withoutPINHandler(withRetries(deviceCallback))
+	_, err = selectDevice(ctx, pin, dev, cb)
 	return trace.Wrap(err)
 }
 
-func findSuitableDevicesOrTimeout(
-	ctx context.Context, filter deviceFilterFunc, knownPaths map[string]struct{}) ([]deviceWithInfo, error) {
-	ticker := time.NewTicker(FIDO2PollInterval)
-	defer ticker.Stop()
-
-	for {
-		switch devices, err := findSuitableDevices(filter, knownPaths); {
-		case err == nil:
-			return devices, nil
-		case errors.Is(err, errNoSuitableDevices):
-			// OK, carry on until we find a device or timeout.
-		default:
-			// Unexpected, abort.
-			return nil, trace.Wrap(err)
-		}
-
-		select {
-		case <-ctx.Done():
-			return nil, trace.Wrap(ctx.Err())
-		case <-ticker.C:
-		}
-	}
-}
-
-func findSuitableDevices(filter deviceFilterFunc, knownPaths map[string]struct{}) ([]deviceWithInfo, error) {
```

Comment:
Most of this function moved to `findDevices` - there are only a few changes in handling `dev.Info()` errors (it's a bit more lenient now).